contrib_bots/lib/github_detail.py

```diff
@@ -0,0 +1,134 @@
+                "If you want to refrence an issues from a repository other than the default use"
+                "the user/repo#id form.")
+
+    def triage_message(self, message, client):
+        # type: () -> bool
+        # Check the message contains a username mention and #<number>
+        # replying to another bot
+        if (re.search("(?:@(?:\*\*)" + re.escape(client.full_name) + ").+?#\d+", message['content']) and
+                not message['sender_email'].endswith('-bot@zulip.com')):
+            return True
+        return False
+
+    def detail_issue(self, issue):
+        # type: (Dict[Text, Union[Text, int, bool]]) -> Text
+
+        id = issue['number']
+        title = issue['title']
+        link = issue['html_url']
+        # Truncate if longer than 200 characters.
+        description = (issue['body'][:200] +
+                       '..') if len(issue['body']) > 202 else issue['body']
+        status = issue['state'].title()
+
+        return '**[{id} | {title}]({link})** - **{status}**\n```quote\n{description}\n```'\
+            .format(id=id, title=title, link=link, status=status, description=description)
+
+    def detail_pull(self, pull):
+        # type: (Dict[Text, Union[Text, int, bool, Dict) -> Text
+
+        id = pull['number']
+        title = pull['title']
+        link = pull['html_url']
+        # Truncate if longer than 200 characters.
+        description = (pull['body'][:200] +
+                       '..') if len(pull['body']) > 202 else pull['body']
+        status = pull['state'].title()
+
+        message = """**[{id} | {title}]({link})** - **{status}**\n```quote\n{description}\n```""".format(
+            id=id, title=title, link=link, status=status, description=description)
+        return message
+
+    def get_details(self, owner, repo, id):
+        # type: (Text, Text, Text) -> Dict[Text, Union[Text, Int, Bool]]
+        # Gets the details of an issues or pull request
+
+        # Try to get an issue, try to get a pull if that fails
+        r = requests.get(
+            'https://api.github.com/repos/{owner}/{repo}/issues/{id}'.format(owner=owner, repo=repo, id=id))
+        if r.status_code == 404:
+            r = requests.get(
+                'https://api.github.com/repos/{owner}/{repo}/pulls/{id}'.format(owner=owner, repo=repo, id=id))
+
+        if r.status_code != requests.codes.ok:
+            return None
+
+        # Check if we have an issue or a pull
+        if re.match('issues\/\d+$', r.url):
+            return self.detail_issue(r.json())
+        else:
+            return self.detail_pull(r.json())
+
+    def handle_message(self, message, client, state_handler):
+        # type: () -> None
+        # Capture owner, repo, id
+        issue_prs = re.finditer(
+            "(?:([\w-]+)?\/)?([\w-]+)?#(\d+)", message['content'])
+
+        bot_messages = []
+        repo = None
+        owner = None
+        for issue_pr in issue_prs:
+            if issue_pr.group(2) != None:
```

Comment:
Create a separate function to extract `repo` and `owner` from the message. All your regexes should be compiled before hand using `re.compile`.